6. Check out the program `dir_tree_gen`.
   1. Make sure it is in the same directory as `data`. Then, execute it (how?). What does it do?
   2. After having finished i.:
       - How many directories does `topdir` contain? 
       - How many `txt` files does `topdir` contain?
       - Which of these `txt` files contain the string `love` most often?
   3. After having finished ii.:
       - Move all `txt` files below `topdir` into `topdir` with a one-line command.
       - Rename all `txt` files in `topdir`now in such a way, that the file count is put between the filename and the extension, separated by `_`. Use zeros as padding, i.e., if there are a 3-digit number of files, the file counts should also be 3-digit. E.g.: `4b17d6.txt` becomes `4b17d6_023.txt`. 
       - Put today's date in the form `YYYYMMDD` at the beginning of each filename, separated by an underscore `_` from the original filename. E.g.: `4b17d6_020.txt` becomes `20240925_4b17d6_020.txt`. *Hint*: `for` loop
       - How can you concatenate all `txt` files into one `txt` file named `concat_shakespeare.txt`? Is it the same as `data/shakespeare.txt`?

In [200]:
cd /tmp
rm -rf terminal_course

In [201]:
mkdir terminal_course
cd terminal_course
ls

In [202]:
cp -r '/Users/omer/Programming/terminal_course/data' .
cp '/Users/omer/Programming/terminal_course/dir_tree_gen' .
ls

data		dir_tree_gen


In [203]:
ls -lh

total 8
drwxr-xr-x@ 3 omer  wheel    96B 24 Dez 03:13 data
-rw-r--r--  1 omer  wheel   842B 24 Dez 03:13 dir_tree_gen


In [204]:
python3 dir_tree_gen
ls

data		dir_tree_gen	topdir


In [205]:
cd topdir

       - How many directories does `topdir` contain? 

In [206]:
find . -type d | wc -l

    3907


       - How many `txt` files does `topdir` contain?

In [207]:
find . -type f -name '*.txt' | wc -l

    7812


       - How many text files contain the word "love"?

In [208]:
find . -type f -name '*.txt' | xargs cat | grep -o 'love' | wc -l

    1278


        - the file names which contain the word love but also show the word

In [209]:
grep -o -r "love" . | head

./197872/a61eb8/0c889e/4e0b3e/0530f7/41ba34/f1e6cb.txt:love
./197872/a61eb8/0c889e/4e0b3e/cc9b43/17a3e0/725826.txt:love
./197872/a61eb8/0c889e/4e0b3e/cc9b43/f0e809.txt:love
./197872/a61eb8/0c889e/4e0b3e/cc9b43/926b9e/c0eac4.txt:love
./197872/a61eb8/0c889e/4e0b3e/cc9b43/f70aae/853ee6.txt:love
./197872/a61eb8/0c889e/8b88af/835136.txt:love
./197872/a61eb8/0c889e/8b88af/ad4d6a/bad89c/4d102d.txt:love
./197872/a61eb8/0c889e/8b88af/c33711/5ae73c.txt:love
./197872/a61eb8/0c889e/8b88af/c33711/5ae73c.txt:love
./197872/a61eb8/0c889e/8b88af/c33711/0480f9/c1419d.txt:love


       - the file names which contain the word "love" without showing the word

In [210]:
grep -o -rl "love" . | head

./197872/a61eb8/0c889e/4e0b3e/0530f7/41ba34/f1e6cb.txt
./197872/a61eb8/0c889e/4e0b3e/cc9b43/17a3e0/725826.txt
./197872/a61eb8/0c889e/4e0b3e/cc9b43/f0e809.txt
./197872/a61eb8/0c889e/4e0b3e/cc9b43/926b9e/c0eac4.txt
./197872/a61eb8/0c889e/4e0b3e/cc9b43/f70aae/853ee6.txt
./197872/a61eb8/0c889e/8b88af/835136.txt
./197872/a61eb8/0c889e/8b88af/ad4d6a/bad89c/4d102d.txt
./197872/a61eb8/0c889e/8b88af/c33711/5ae73c.txt
./197872/a61eb8/0c889e/8b88af/c33711/0480f9/c1419d.txt
./197872/a61eb8/0c889e/8b88af/c33711/7e77ed/31521f.txt


       - Which of these `txt` files contain the string `love` most often?

In [211]:
max_count=0;
max_file="";
for file in $(grep -o -rl "love" .); do 
    count=$(grep -o "love" $file | wc -l);
    if (( count > max_count )); then
        max_count=$count;
        max_file=$file;
    fi;
done;
echo $max_count
echo $max_file

3
./197872/a61eb8/4602bf/d25f9d/433c41/50bbea/af995b.txt


In [212]:
pwd

/tmp/terminal_course/topdir


       - Move all `txt` files below `topdir` into `topdir` with a one-line command.

In [213]:
find . -type f -name "*.txt" -exec mv {} . \;
ls | wc -l

    7812


       - Rename all `txt` files in `topdir`now in such a way, that the file count is put between the filename and the extension, separated by `_`. Use zeros as padding, i.e., if there are a 3-digit number of files, the file counts should also be 3-digit. E.g.: `4b17d6.txt` becomes `4b17d6_023.txt`.

In [214]:
ls | head

0009a1.txt
00101d.txt
0017c4.txt
001954.txt
001acf.txt
00320b.txt
00380a.txt
003de0.txt
00479f.txt
004c5b.txt


In [215]:
counter=0; 
for file in *.txt; do 
    printf -v padding "%04d" $counter; 
    mv $file "${file/.txt/_${padding}.txt}"; 
    counter=$((counter + 1));
done;
ls | head

0009a1_0000.txt
00101d_0001.txt
0017c4_0002.txt
001954_0003.txt
001acf_0004.txt
00320b_0005.txt
00380a_0006.txt
003de0_0007.txt
00479f_0008.txt
004c5b_0009.txt


In [216]:
ls | tail

ffa2eb_7801.txt
ffaac4_7802.txt
ffafa8_7803.txt
ffb351_7804.txt
ffb5d3_7805.txt
ffc0cf_7806.txt
ffcb11_7807.txt
ffcece_7808.txt
ffdf12_7809.txt
ffe03f_7810.txt


       - Put today's date in the form `YYYYMMDD` at the beginning of each filename, separated by an underscore `_` from the original filename. E.g.: `4b17d6_020.txt` becomes `20240925_4b17d6_020.txt`. *Hint*: `for` loop

In [217]:
date=$(date "+%Y%m%d")
echo $date

20241224


In [218]:
for file in *.txt; do mv $file "${date}_${file}"; done;
ls | head

197872
20241224_0009a1_0000.txt
20241224_00101d_0001.txt
20241224_0017c4_0002.txt
20241224_001954_0003.txt
20241224_001acf_0004.txt
20241224_00320b_0005.txt
20241224_00380a_0006.txt
20241224_003de0_0007.txt
20241224_00479f_0008.txt


       - How can you concatenate all `txt` files into one `txt` file named `concat_shakespeare.txt`? Is it the same as `data/shakespeare.txt`?

In [219]:
for file in *.txt; do cat $file >> concat_shakespeare.txt; done;
cat concat_shakespeare.txt | head

And patience tame to sufferance bide each check,
Without accusing you of injury.  
For since each hand hath put on nature's power,  
Fairing the foul with art's false borrowed face,
Points on me graciously with fair aspect,
And puts apparel on my tattered loving,
Will sourly leave her till he have prevailed?
Ay me, but yet thou mightst my seat forbear,
And lace it self with his society?  
Why should false painting imitate his cheek,


In [220]:
cat /tmp/terminal_course/data/shakespeare.txt | head

THE SONNETS

by William Shakespeare

From fairest creatures we desire increase,
That thereby beauty's rose might never die,
But as the riper should by time decease,
His tender heir might bear his memory:
But thou contracted to thine own bright eyes,
Feed'st thy light's flame with self-substantial fuel,


In [222]:
cat concat_shakespeare.txt | wc -l

   14806


In [223]:
cat /tmp/terminal_course/data/shakespeare.txt | wc -l

    2468


No they are NOT the same since one file contains 14806 line and the other contains only 2468 lines and the Headers are also different.

In [224]:
rm -rf /tmp/terminal_course